src/main/java/org/dependencytrack/persistence/VulnerableSoftwareQueryManager.java

```diff
@@ -127,9 +129,37 @@ public VulnerableSoftware getVulnerableSoftwareByCpe23(String cpe23,
                                                            String versionEndExcluding, String versionEndIncluding,
                                                            String versionStartExcluding, String versionStartIncluding) {
         final Query<VulnerableSoftware> query = pm.newQuery(VulnerableSoftware.class);
-        query.setFilter("cpe23 == :cpe23 && versionEndExcluding == :versionEndExcluding && versionEndIncluding == :versionEndIncluding && versionStartExcluding == :versionStartExcluding && versionStartIncluding == :versionStartIncluding");
+        var filter = "cpe23 == :cpe23";
+        final var parameters = new HashMap<String, Object>();
+        parameters.put("cpe23", Objects.requireNonNull(cpe23));
+        if (versionEndExcluding != null) {
+            filter += " && versionEndExcluding == :vee";
+            parameters.put("vee", versionEndExcluding);
+        } else {
+            filter += " && versionEndExcluding == null";
+        }
+        if (versionEndIncluding != null) {
+            filter += " && versionEndIncluding == :vei";
+            parameters.put("vei", versionEndIncluding);
+        } else {
+            filter += " && versionEndIncluding == null";
+        }
+        if (versionStartExcluding != null) {
+            filter += " && versionStartExcluding == :vse";
+            parameters.put("vse", versionStartExcluding);
+        } else {
+            filter += " && versionStartExcluding == null";
+        }
+        if (versionStartIncluding != null) {
+            filter += " && versionStartIncluding == :vsi";
+            parameters.put("vsi", versionStartIncluding);
+        } else {
+            filter += " && versionStartIncluding == null";
+        }
+        query.setFilter(filter);
+        query.setNamedParameters(parameters);
```

Comment:
Should a comment be added to explain why the query is built up like this, I assume it's because of that null parameters breaking query compilation caching?

Reply:
Yes, good point. I will add a comment.